StripeCore/StripeCore/Source/Attestation/StripeAttest.swift

```diff
@@ -0,0 +1,278 @@
+            throw AttestationError.attestationNotSupported
+        }
+        if let keyId = Self.keyID {
+            return keyId
+        }
+        // If we don't have a key, generate one.
+        return try await self.createKey()
+    }
+
+    @_spi(STP) public func resetKey() {
+        Self.keyID = nil
+        Self.successfullyAttested = false
+    }
+
+    private func createKey() async throws -> String {
+        let keyId = try await appAttestService.generateKey()
+        // Save the Key ID, and that the key is not attested.
+        Self.keyID = keyId
+        Self.successfullyAttested = false
+        return keyId
+    }
+
+    func getAppID() throws -> String {
+        if let appID = Bundle.main.bundleIdentifier {
+            return appID
+        }
+        throw AttestationError.noAppID
+    }
+
+    func getDeviceID() throws -> String {
+        if let deviceID = UIDevice.current.identifierForVendor?.uuidString {
+            return deviceID
+        }
+        throw AttestationError.noDeviceID
+    }
+
+    /// Get a challenge from the backend.
+    func getChallenge() async throws -> String {
+        let keyID = try await self.getOrCreateKeyID()
+        return try await appAttestBackend.getChallenge(appId: getAppID(), deviceId: getDeviceID(), keyId: keyID)
+    }
+
+    /// Generate the assertion data from a key and challenge.
+    private func generateAssertion(keyId: String, challenge: String, retryIfNeeded: Bool = true) async throws -> Data {
+        // We're just signing the challenge for now.
+        // The expected format is the SHA256 hash of the JSON-encoded dictionary.
+        let assertionDictionary = [ "challenge": challenge ]
+        let assertionData = try JSONSerialization.data(withJSONObject: assertionDictionary,
+                                                       // Sort our keys: It's important that the JSON is
+                                                       // the same on the backend and frontend!
+                                                       options: [.sortedKeys])
+        let assertionDataHash = Data(SHA256.hash(data: assertionData))
+        do {
+            return try await appAttestService.generateAssertion(keyId, clientDataHash: assertionDataHash)
+        } catch {
+            // If error is DCErrorInvalidKey (3) and the domain is DCErrorDomain,
+            // then the key is either unattested or corrupted.
+            let error = error as NSError
+            if error.domain == DCErrorDomain && error.code == DCError.invalidKey.rawValue {
+                if retryIfNeeded {
+                    // We'll try to attest again, maybe our initial attestation was unsuccessful?
+                    // `DCError.invalidKey` could mean a lot of things, unfortunately.
+                    // If this doesn't work, then in `attest()` we'll deem the key to be corrupted
+                    // and throw it out.
+                    try await attest()
```

Comment:
It isn't a straightforward retry, it's re-attesting the key (in case it wasn't correctly attested) and *then* retrying with the re-attested key or a new key. So a timeout wouldn't change the behavior. I can make that clearer in the comments/naming!

Reply:
Ahh gotcha, that makes sense!